### PR TITLE
Add retries to get head_pipeline before attempting a merge

### LINKED ISSUE #16: Auto-merge does not work sometimes
We observe this in our private PackageRegistry sometimes. 

Registrator/TagBot will open the PR, a `registrar` GitLab user will approve the PR, but then sometimes _not_ enabled an automatic merge when the pipeline succeeds. 

Since users take for granted that registry PRs will be auto-merged, this can often lead to PRs being left often (and versions not being registered) until someone notices and merges the PR manually.

It is relatively rarely... maybe 1% of the time at a rough guess.

### PR DESCRIPTION
**Pull request by nicoleepp**
Might close #16

### CODE REVIEW

File: tagbotgitlab/tagbot.py [hunk 1]
```diff
@@ -65,13 +66,25 @@ def handle_open(payload):
...
     mr = p.mergerequests.get(mr_id, lazy=False)
     print(mr)
-    print("Merging MR")
+
+    # Add timeout to wait for the head pipeline to be associated properly with the MR
+    # Avoids merge failures
+    timeout = 1
+    while timeout <= 3 and mr.head_pipeline is None:
+        print(f"Sleeping for {timeout} seconds")
+        time.sleep(timeout)
```
Comment (fchorney) [inline tagbotgitlab/tagbot.py:83]: This is going to sleep for 1 second, then 2, then 3 was this intended? I guess its kind of like an exponential timeout type of thing?

File: tagbotgitlab/tagbot.py [hunk 2]
```diff
@@ -65,13 +66,25 @@ def handle_open(payload):
...
     # returns an error
     mr = p.mergerequests.get(mr_id, lazy=False)
     print(mr)
-    print("Merging MR")
+
+    # Add timeout to wait for the head pipeline to be associated properly with the MR
+    # Avoids merge failures
+    timeout = 1
+    while timeout <= 3 and mr.head_pipeline is None:
+        print(f"Sleeping for {timeout} seconds")
```
Comment (fchorney) [inline tagbotgitlab/tagbot.py:82]: Maybe something like this so it's clear why it's sleeping
```suggestion
        print(f"head_pipeline not set - sleeping for {timeout} seconds")
```

File: tests/test_tagbot.py [hunk 3]
```diff
@@ -267,6 +268,38 @@ def test_handle_open():
...
+    p.mergerequests.get.assert_has_calls(calls)
+    mr.approve.assert_called_once_with()
+    mr.merge.assert_called_once_with(
+        merge_when_pipeline_succeeds=True, should_remove_source_branch=True
+    )
+
+    # Test with head pipeline not set
+    # Note it should still approve and try merging, GitLab API will throw an error if
+    # the merge is not valid but this is mocked so it succeeds fine
+    mr = Mock(spec=gitlab.v4.objects.ProjectMergeRequest)
```
Comment (fchorney) [inline tests/test_tagbot.py:282]: I'm not super familiar with `Mock`, does the `sleep` still activate when this stuff is mocked? you could perhaps mock `sleep` so it doesn't actually take 1 second (if it does)

File: tagbotgitlab/tagbot.py (hunk 1 above)
Comment (nicoleepp) [reply tagbotgitlab/tagbot.py:83]: Yeah, it was intended to work as a bootleg exponential backoff

File: tests/test_tagbot.py (hunk 3 above)
Comment (nicoleepp) [reply tests/test_tagbot.py:283]: Thanks for the tip

**Review by fchorney — APPROVED** [review]

**Pull request merged**